google/resource_kms_crypto_key.go

```diff
@@ -69,179 +96,152 @@ func resourceKmsCryptoKey() *schema.Resource {
-
-	cryptoKey, err := config.clientKms.Projects.Locations.KeyRings.CryptoKeys.Get(cryptoKeyId.cryptoKeyId()).Do()
+	res, err := sendRequest(config, "GET", url, nil)
 	if err != nil {
-		return fmt.Errorf("Error reading CryptoKey: %s", err)
+		return handleNotFoundError(err, d, fmt.Sprintf("KmsCryptoKey %q", d.Id()))
 	}
-	d.Set("key_ring", cryptoKeyId.KeyRingId.terraformId())
-	d.Set("name", cryptoKeyId.Name)
-	d.Set("rotation_period", cryptoKey.RotationPeriod)
-	d.Set("self_link", cryptoKey.Name)
 
-	if err = d.Set("version_template", flattenVersionTemplate(cryptoKey.VersionTemplate)); err != nil {
-		return fmt.Errorf("Error setting version_template in state: %s", err.Error())
+	res, err = resourceKmsCryptoKeyDecoder(d, meta, res)
+	if err != nil {
+		return err
 	}
 
-	d.SetId(cryptoKeyId.cryptoKeyId())
+	if err := d.Set("purpose", flattenKmsCryptoKeyPurpose(res["purpose"], d)); err != nil {
+		return fmt.Errorf("Error reading CryptoKey: %s", err)
+	}
+	if err := d.Set("rotation_period", flattenKmsCryptoKeyRotationPeriod(res["rotationPeriod"], d)); err != nil {
+		return fmt.Errorf("Error reading CryptoKey: %s", err)
+	}
+	if err := d.Set("version_template", flattenKmsCryptoKeyVersionTemplate(res["versionTemplate"], d)); err != nil {
+		return fmt.Errorf("Error reading CryptoKey: %s", err)
+	}
 
 	return nil
 }
 
-func clearCryptoKeyVersions(cryptoKeyId *kmsCryptoKeyId, config *Config) error {
-	versionsClient := config.clientKms.Projects.Locations.KeyRings.CryptoKeys.CryptoKeyVersions
+func resourceKmsCryptoKeyUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
 
-	versionsResponse, err := versionsClient.List(cryptoKeyId.cryptoKeyId()).Do()
+	obj := make(map[string]interface{})
+	rotationPeriodProp, err := expandKmsCryptoKeyRotationPeriod(d.Get("rotation_period"), d, config)
+	if err != nil {
+		return err
+	} else if v, ok := d.GetOkExists("rotation_period"); !isEmptyValue(reflect.ValueOf(v)) && (ok || !reflect.DeepEqual(v, rotationPeriodProp)) {
+		obj["rotationPeriod"] = rotationPeriodProp
+	}
+	versionTemplateProp, err := expandKmsCryptoKeyVersionTemplate(d.Get("version_template"), d, config)
+	if err != nil {
+		return err
+	} else if v, ok := d.GetOkExists("version_template"); !isEmptyValue(reflect.ValueOf(v)) && (ok || !reflect.DeepEqual(v, versionTemplateProp)) {
+		obj["versionTemplate"] = versionTemplateProp
+	}
 
+	obj, err = resourceKmsCryptoKeyUpdateEncoder(d, meta, obj)
 	if err != nil {
 		return err
 	}
 
-	for _, version := range versionsResponse.CryptoKeyVersions {
-		request := &cloudkms.DestroyCryptoKeyVersionRequest{}
-		_, err = versionsClient.Destroy(version.Name, request).Do()
+	url, err := replaceVars(d, config, "https://cloudkms.googleapis.com/v1/{{key_ring}}/cryptoKeys/{{name}}")
+	if err != nil {
+		return err
+	}
 
-		if err != nil {
-			return err
-		}
+	log.Printf("[DEBUG] Updating CryptoKey %q: %#v", d.Id(), obj)
+	updateMask := []string{}
+
+	if d.HasChange("rotation_period") {
```

Comment:
Because `nextRotationTime` is coupled to `rotationPeriod` inside the update encoder, we need to include it in the `updateMask` when updating the later. Skipping it makes the `nextRotationTime` to not be updated when `rotationPeriod` gets updated, and even worse, causing an error of "`nextRotationTime` not specified" when you define the `rotationPeriod` for the first time through an update.

Any way we can influence which fields get included in the `updateMask`?

Reply:
Not at the moment, but you're welcome to add it! There's a comment in the file that explains what's currently supported and a possible fix: https://github.com/GoogleCloudPlatform/magic-modules/blob/master/templates/terraform/update_mask.erb

Reply:
You could also do it with custom code, though I'd rather have the real fix if that's possible

Reply:
Not sure if that's what you had in mind, but having a `update_mask_fields` seems flexible and simple.